File: 220_Contains_Duplicate_III_Bucket.java
public class Solution {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int k, int t) {
        if (nums.length == 0 || t < 0)  return false;
        int rmIdx = 0;
        Map<Integer, Integer> buckets = new HashMap();
        for (int i = 0; i < nums.length; i ++){
            int toAdd = nums[i];
            int bucketNum = getBucketNum(toAdd, t+1);
            if (buckets.containsKey(bucketNum))    return true;
            if (buckets.containsKey(bucketNum+1) && (long) buckets.get(bucketNum+1) - toAdd <= t){
                return true;
            }if (buckets.containsKey(bucketNum-1) && (long) toAdd - buckets.get(bucketNum-1) <= t){
                return true;
            }
            buckets.put(bucketNum, toAdd);
            if (buckets.size() > k){
                buckets.remove(nums[rmIdx++] / (t + 1));
            }
        }
        return false;
    }
    public int getBucketNum(int num, int base){
        return num >= 0 ? num / base: (num+1) / base - 1;
    }
}

File: python/220_Contains_Duplicate_III_Bucket.py
# Given an array of integers, find out whether there are two distinct indices 
# i and j in the array such that the absolute difference between nums[i] and nums[j]
# is at most t and the absolute difference between i and j is at most k.
# 

class Solution(object):
    def containsNearbyAlmostDuplicate(self, nums, k, t):
        """
        :type nums: List[int]
        :type k: int
        :type t: int
        :rtype: bool
        """
        if not nums:
            return False
        if t < 0:
            return False
        buckets = {}
        rmIdx = 0
        bucketBase = t + 1
        for num in nums:
            bucketNum = num / bucketBase
            if bucketNum in buckets:
                return True
            if bucketNum + 1 in buckets and buckets[bucketNum + 1] - num <= t:
                return True
            if bucketNum - 1 in buckets and num - buckets[bucketNum - 1] <= t:
                return True
            buckets[bucketNum] = num
            if len(buckets) > k:
                buckets.pop(nums[rmIdx] / bucketBase)
                rmIdx += 1
        return Falsehttps://leetcode.com/submissions/detail/93116538/   